### DISCUSSION TITLE
support for union with different number of columns

### DISCUSSION THREAD

**Discussion by djouallah** [duckdb/duckdb#4308] [Ideas]
fellow this
https://github.com/duckdb/duckdb/issues/4302

it will be nice if somehow DuckDB accept the union of two tables with different number of columns by default, it is very common when doing transformation

**Comment by Mytherin** [comment]
Thanks for the suggestion! Could you perhaps expand with some examples of what exactly you want to achieve, and how you would do so in other libraries (e.g. Pandas) or how you would write a SQL query currently that solves the problem? Do the tables have “matching” columns that have the same name that should be overlapped?

**Comment by djouallah** [comment]
yes  the tables have matching names, except some tables may have extra columns, we expected those columns to show up with null values where it is not available,
we have this a lot at work, we use PowerQuery ( proprietary Microsoft ETL tools) which handle it by default, it is called append mode

![image](https://user-images.githubusercontent.com/12554469/183243068-2739ada5-26fd-4990-859f-d1f054305d17.png)

**Comment by Alex-Monahan** [comment]
I think there are 2 potential feature additions here! 
* A union with different numbers of columns
* A union by column name instead of column position

Both are supported in a Pandas.concat operation, and by Spark here: https://spark.apache.org/docs/3.1.1/api/python/reference/api/pyspark.sql.DataFrame.unionByName.html

**Reply by Mytherin** [reply]
Agreed, I think these are related as well. I am just wondering about the syntax. Perhaps “UNION BY NAME”? Happy to hear other more elegant suggestions as well.

**Reply by Alex-Monahan** [reply]
If the parser can handle UNION BY NAME, that is quite nice! Another option is CONCAT, but we already have group_concat as an agg function.

**Comment by djouallah** [comment]
in PowerQuery we use the verb Append :) I think it is not bad

**Comment by danthegoodman1** [comment]
I think this would be particularly helpful when you are doing something like `select count(distinct b) from read_parquet('*.parquet')` where `a.parquet` does not have the `b` column, but `b.parquet` does. Accounts for an evolving parquet schema

**Comment by simonaubertbd** [comment]
I totally support this idea, it's very common in data to create "fact tables" with several kind of facts (like orders and invoices), each one having different columns. Some tools already accepts that (like Qlik with the concatenate function or Alteryx with the union tool).

But maybe it shouldn't be a function called "union". union_something would be better. is there anything like that in other bases or in sql norm we can relate to?

Best regards,

Simon

**Comment by djouallah** [comment]
@simonaubertbd it is supported already just use union with name (check the documentation)

**Reply by simonaubertbd** [reply]
Oh, sorry. Cool.
